<center> Andrea Elliott <center> <br> 
<center> MSDE 620: Week 5 <center> <br>
<center> Due: September 25, 2022 <center>


Here I focus on cleaning and preparing a dataset that looks at adult employment. First I will load the data and visually examine it.

In [88]:
import pandas as pd
import numpy as np

In [1]:
!head adult.data

39, State-gov, 77516, Bachelors, 13, Never-married, Adm-clerical, Not-in-family, White, Male, 2174, 0, 40, United-States, <=50K
50, Self-emp-not-inc, 83311, Bachelors, 13, Married-civ-spouse, Exec-managerial, Husband, White, Male, 0, 0, 13, United-States, <=50K
38, Private, 215646, HS-grad, 9, Divorced, Handlers-cleaners, Not-in-family, White, Male, 0, 0, 40, United-States, <=50K
53, Private, 234721, 11th, 7, Married-civ-spouse, Handlers-cleaners, Husband, Black, Male, 0, 0, 40, United-States, <=50K
28, Private, 338409, Bachelors, 13, Married-civ-spouse, Prof-specialty, Wife, Black, Female, 0, 0, 40, Cuba, <=50K
37, Private, 284582, Masters, 14, Married-civ-spouse, Exec-managerial, Wife, White, Female, 0, 0, 40, United-States, <=50K
49, Private, 160187, 9th, 5, Married-spouse-absent, Other-service, Not-in-family, Black, Female, 0, 0, 16, Jamaica, <=50K
52, Self-emp-not-inc, 209642, HS-grad, 9, Married-civ-spouse, Exec-managerial, Husband, White, Male, 0, 0, 45, United-States, >50K
31, 

There's no header row. Column names can be generated from the names file.

In [271]:
col_names = ['Age', 'Work Class', 'Final Weight', 'Education', 'Education Number', 'Marital Status',
           'Occupation', 'Relationship', 'Race', 'Sex', 'Capital Gain', 'Capital Loss', 'Hours per Week', 
           'Native Country', 'Income Prediction']

In [272]:
birth = pd.read_csv('adult.data', names=col_names)

In [273]:
birth.head(10)

Age         Work Class  Final Weight   Education  Education Number  \
0   39          State-gov         77516   Bachelors                13   
1   50   Self-emp-not-inc         83311   Bachelors                13   
2   38            Private        215646     HS-grad                 9   
3   53            Private        234721        11th                 7   
4   28            Private        338409   Bachelors                13   
5   37            Private        284582     Masters                14   
6   49            Private        160187         9th                 5   
7   52   Self-emp-not-inc        209642     HS-grad                 9   
8   31            Private         45781     Masters                14   
9   42            Private        159449   Bachelors                13   

           Marital Status          Occupation    Relationship    Race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black   
4      Married-civ-spouse      Prof-specialty            Wife   Black   
5      Married-civ-spouse     Exec-managerial            Wife   White   
6   Married-spouse-absent       Other-service   Not-in-family   Black   
7      Married-civ-spouse     Exec-managerial         Husband   White   
8           Never-married      Prof-specialty   Not-in-family   White   
9      Married-civ-spouse     Exec-managerial         Husband   White   

       Sex  Capital Gain  Capital Loss  Hours per Week  Native Country  \
0     Male          2174             0              40   United-States   
1     Male             0             0              13   United-States   
2     Male             0             0              40   United-States   
3     Male             0             0              40   United-States   
4   Female             0             0              40            Cuba   
5   Female             0             0              40   United-States   
6   Female             0             0              16         Jamaica   
7     Male             0             0              45   United-States   
8   Female         14084             0              50   United-States   
9     Male          5178             0              40   United-States   

  Income Prediction  
0             <=50K  
1             <=50K  
2             <=50K  
3             <=50K  
4             <=50K  
5             <=50K  
6             <=50K  
7              >50K  
8              >50K  
9              >50K

#### Replace '-' with spaces <br>

As I glance through the header above, it appears there are many hyphens in the string entries. The apply function will be appropriate here. I will write a function to replace these hyphens with spaces. I only want to work with strings, which means I'm looking specifically for the 'object' columns


In [274]:
birth.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Age                32561 non-null  int64 
 1   Work Class         32561 non-null  object
 2   Final Weight       32561 non-null  int64 
 3   Education          32561 non-null  object
 4   Education Number   32561 non-null  int64 
 5   Marital Status     32561 non-null  object
 6   Occupation         32561 non-null  object
 7   Relationship       32561 non-null  object
 8   Race               32561 non-null  object
 9   Sex                32561 non-null  object
 10  Capital Gain       32561 non-null  int64 
 11  Capital Loss       32561 non-null  int64 
 12  Hours per Week     32561 non-null  int64 
 13  Native Country     32561 non-null  object
 14  Income Prediction  32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [132]:
birth_new = birth.apply(lambda x: x.str.replace('-', ' ') if x.dtype == "O" else x)

In [134]:
birth_new.head()

Age         Work Class  Final Weight   Education  Education Number  \
0   39          State gov         77516   Bachelors                13   
1   50   Self emp not inc         83311   Bachelors                13   
2   38            Private        215646     HS grad                 9   
3   53            Private        234721        11th                 7   
4   28            Private        338409   Bachelors                13   

        Marital Status          Occupation    Relationship    Race      Sex  \
0        Never married        Adm clerical   Not in family   White     Male   
1   Married civ spouse     Exec managerial         Husband   White     Male   
2             Divorced   Handlers cleaners   Not in family   White     Male   
3   Married civ spouse   Handlers cleaners         Husband   Black     Male   
4   Married civ spouse      Prof specialty            Wife   Black   Female   

   Capital Gain  Capital Loss  Hours per Week  Native Country  \
0          2174             0              40   United States   
1             0             0              13   United States   
2             0             0              40   United States   
3             0             0              40   United States   
4             0             0              40            Cuba   

  Income Prediction  
0             <=50K  
1             <=50K  
2             <=50K  
3             <=50K  
4             <=50K

#### Determine how to deal with missing values (if any)  <br>
The first step here is just to see if any values are missing.

In [135]:
birth.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Age                32561 non-null  int64 
 1   Work Class         32561 non-null  object
 2   Final Weight       32561 non-null  int64 
 3   Education          32561 non-null  object
 4   Education Number   32561 non-null  int64 
 5   Marital Status     32561 non-null  object
 6   Occupation         32561 non-null  object
 7   Relationship       32561 non-null  object
 8   Race               32561 non-null  object
 9   Sex                32561 non-null  object
 10  Capital Gain       32561 non-null  int64 
 11  Capital Loss       32561 non-null  int64 
 12  Hours per Week     32561 non-null  int64 
 13  Native Country     32561 non-null  object
 14  Income Prediction  32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


#### Use apply() with User Defined Functions (UDFs) to analyze missing values

## def count_missing(vec):
    """counts the number of missing vales"""
    null_vec = pd.isnull(vec)
    null_count = np.sum(null_vec)
    return null_count

In [276]:
col_missing_count = birth.apply(count_missing)
col_missing_count

Age                  0
Work Class           0
Final Weight         0
Education            0
Education Number     0
Marital Status       0
Occupation           0
Relationship         0
Race                 0
Sex                  0
Capital Gain         0
Capital Loss         0
Hours per Week       0
Native Country       0
Income Prediction    0
dtype: int64

It looks like we got a complete dataset this time. However, sometimes entries like 'unknown' and other strings aren't recognized as missing values. I will try strings like '?' and 'unknown' in the code below to try to identify any missing values.

In [277]:
birth.eq(' ?').sum()

Age                     0
Work Class           1836
Final Weight            0
Education               0
Education Number        0
Marital Status          0
Occupation           1843
Relationship            0
Race                    0
Sex                     0
Capital Gain            0
Capital Loss            0
Hours per Week          0
Native Country        583
Income Prediction       0
dtype: int64

Aha! If I place a space before that question mark, I can identify some missing values. Let's first look at the entries for those columns.

In [280]:
birth.groupby('Work Class').size()

Work Class
 ?                    1836
 Federal-gov           960
 Local-gov            2093
 Never-worked            7
 Private             22696
 Self-emp-inc         1116
 Self-emp-not-inc     2541
 State-gov            1298
 Without-pay            14
dtype: int64

In this situation, over half of the entries are in the private sector. Although I would be comfortable leaving the ?s as their own category, I will replace them using the mode with a UDF (below, after I decide what to do with the two other columns containing missing values). <br>

Now I will look at the Occupation column, which is also missing data.

In [202]:
birth.groupby('Occupation').size()

Occupation
 ?                    1843
 Adm-clerical         3770
 Armed-Forces            9
 Craft-repair         4099
 Exec-managerial      4066
 Farming-fishing       994
 Handlers-cleaners    1370
 Machine-op-inspct    2002
 Other-service        3295
 Priv-house-serv       149
 Prof-specialty       4140
 Protective-serv       649
 Sales                3650
 Tech-support          928
 Transport-moving     1597
dtype: int64

Values for these are more evenly split across different categories for the 'Occupation' column. Since the most common occupation ('Prof-specialty') has less than 100 more entries than the next two most common categories, I don't think it would be a good idea to just replace the '?' entries with the mode ('Prof-specialty'). Doing so would lead to a nearly 50% increase in the mode, making it by far the most common entry, when at this point, other entries are not far behind. For 'Occupation', I will leave the unknown as its own category. Interestingly, there is no option of 'unemployed' or 'stay at home parent'. It's possible that those ?s could be at least in part unemployed individuals.

In [282]:
birth.groupby('Native Country').size()

Native Country
 ?                               583
 Cambodia                         19
 Canada                          121
 China                            75
 Columbia                         59
 Cuba                             95
 Dominican-Republic               70
 Ecuador                          28
 El-Salvador                     106
 England                          90
 France                           29
 Germany                         137
 Greece                           29
 Guatemala                        64
 Haiti                            44
 Holand-Netherlands                1
 Honduras                         13
 Hong                             20
 Hungary                          13
 India                           100
 Iran                             43
 Ireland                          24
 Italy                            73
 Jamaica                          81
 Japan                            62
 Laos                             18
 Mexico                

While these ?s may be listed as such for a specific reason, I see that the vast majority of entries in the dataset are from people born in the United States. I will replace these ?s with United States as the native country. <br>

Now to actually replace those values in the 'Native Country' and 'Work Class' columns.

In [315]:
def replace_unk(col):
    col = col.replace(' ?', col.mode()[0])
    return col

In [321]:
birth['Native Country'] = replace_unk(birth['Native Country'])

In [323]:
birth['Work Class'] = replace_unk(birth['Work Class'])

In [324]:
birth.eq(' ?').sum()

Age                     0
Work Class              0
Final Weight            0
Education               0
Education Number        0
Marital Status          0
Occupation           1843
Relationship            0
Race                    0
Sex                     0
Capital Gain            0
Capital Loss            0
Hours per Week          0
Native Country          0
Income Prediction       0
dtype: int64

It looks like the ?s have been taken care of for 'Work Class' and 'Native Country', and I decided to leave them as their own category for 'Occupation'.

### Use the grouping and aggregation methods, analyze data vs. income several different ways.<br>

#### Age vs. Income
This is when it starts to get fun. First, I'd like to compare the ages of those making less than 50K per year and those making more than 50K. I'd expect those in the higher earner group would have a higher mean age. Before doing that analysis, however, I'm going to add an underscore to the column names to make it easier to structure the code for the aggregate functions. <br>


In [174]:
cols = birth.columns
cols = cols.str.replace(' ', '_')
birth.columns = cols

In [236]:
income_by_age = birth.groupby('Income_Prediction').Age.agg([np.mean, np.std])
income_by_age

mean        std
Income_Prediction                      
 <=50K             36.783738  14.020088
 >50K              44.249841  10.519028

This is consistent with my prediction, although the difference between them is actually larger than I'd expect. This is a rather large standard deviation given the size of the dataset (32561 entries). Therefore, while there is a difference in the age of the two income levels, I would not expect this to be a statistically significant difference, although further analysis would be required to say for sure. <br>

#### Work Class vs. Income

I'd like to find out what the most common Work Class categories are for the two different income levels. Below I show the aggregate function I'd like to use, but I'm not getting a division by Income Prediction as I would expect. It may be because both columns are categorical data. In the second block of code, I use a different method to pull out this information using `groupby` but not `agg()`.

In [215]:
income_by_class = birth.groupby('Work_Class')['Income_Prediction'].agg([np.count_nonzero])
income_by_class

count_nonzero
Work_Class                      
 Federal-gov                 960
 Local-gov                  2093
 Never-worked                  7
 Private                   24532
 Self-emp-inc               1116
 Self-emp-not-inc           2541
 State-gov                  1298
 Without-pay                  14

In [218]:
income_by_class = birth.groupby(['Income_Prediction', 'Work_Class'])
income_by_class.size()

Income_Prediction  Work_Class       
 <=50K              Federal-gov           589
                    Local-gov            1476
                    Never-worked            7
                    Private             19378
                    Self-emp-inc          494
                    Self-emp-not-inc     1817
                    State-gov             945
                    Without-pay            14
 >50K               Federal-gov           371
                    Local-gov             617
                    Private              5154
                    Self-emp-inc          622
                    Self-emp-not-inc      724
                    State-gov             353
dtype: int64

I suppose it's not surprising that both income categories have Private sector as their most common Work Class, as that was far and away the most common Work Class in the data set. It's clear that there are more <=50K entries, but how many more is unclear. Let me pull that out so I can get a better sense of how the percentage of people in the private sector compares across the two income groups. I'm using 'Age' in this aggregate function not because I care about Age but because all the columns should produce the same nonzero count, and I only need one column.

In [221]:
birth.groupby('Income_Prediction').Age.agg(np.count_nonzero)

Income_Prediction
 <=50K    24720
 >50K      7841
Name: Age, dtype: int64

Now I'll look at what percentage of each income group works in the private sector.

In [224]:
print(19378/24720*100, 'percent of people earning equal to or less than 50K work in the private sector')
print(5154/7841*100, 'percent of people earning more than 50K work in the private sector')


78.38996763754047 percent of people earning equal to or less than 50K work in the private sector
65.73141180971814 percent of people earning more than 50K work in the private sector


It looks like a slightly greater percentage of people in the lower income category work in the private sector compared to the higher income category, but in both income categories, the majority of people work in the private sector. <br>

#### Hours per week vs. Income <br>

Are hours related to income?

In [231]:
income_by_hours = birth.groupby('Income_Prediction')['Hours_per_Week'].agg([np.mean])
income_by_hours

mean
Income_Prediction           
 <=50K             38.840210
 >50K              45.473026

Those in the higher income bracket are reporting more working hours per weeek. What are the maximum and minimum number of hours people are reporting at each income level?

In [238]:
max_min_hours = birth.groupby('Income_Prediction')['Hours_per_Week'].agg([np.max, np.min])
max_min_hours

amax  amin
Income_Prediction            
 <=50K               99     1
 >50K                99     1

In [245]:
max_min_hours_all = birth.groupby('Hours_per_Week')['Age'].agg([np.max, np.min, np.count_nonzero])
max_min_hours_all

amax  amin  count_nonzero
Hours_per_Week                           
1                 81    21             20
2                 87    18             32
3                 85    18             39
4                 90    17             54
5                 82    17             60
...              ...   ...            ...
95                52    25              2
96                46    25              5
97                31    29              2
98                63    21             11
99                90    19             85

[94 rows x 3 columns]

Why may those in the higher income category work more hours per week? Perhaps the people in the lower income category tend to be in hourly jobs and therefore don't exceed 40 hours per week, while the people in the higher income category are salaried and are more likely to work extra hours. Alternatively, it could be that those people in the higher income back who are working hourly jobs may be putting in overtime to reach that higher income level. <br>

The second and third aggregate function I ran under this subcategory was looking at the minimum and maximum entries. I found that regardless of income level, the minimum and maximum reported hours of work are 1 and 99, respectively. That led me to look at minimum and maximum age as a function of Hours per Week, as well as the number of people who reported each value for Hours per Week. Above we only see the lowest and highest reports, but that's fine with me because I think those are interesting to consider. First, we're seeing that the maximum ages for those working 1-5 hours per week are 80s and 90s. This make sense, as elderly people may be working just a few hours for something to do. At the same time, I think a healthy skepticism is warranted for those ~90 people reporting 1-3 hours per week. Why would a business keep that 21 year old on payroll for 1 hour per week? Did that person just fill in 1 hour because there was no option for 0 hours? I am also skeptical of those reporting >96 hours per week, which would be about 14 hours of work seven days per week. Is it possible? yes. But is it likely for any length of time? No. However, it's also possible that people are counting other things in their hours of work (e.g., household obligations, etc). What about that 90 year old reporting 99 hours of work per week? I would make a strong claim that that isn't a valid data point.